src/org/traccar/protocol/EelinkProtocolDecoder.java

```diff
@@ -241,45 +235,41 @@ private Position decodeOld(
                 break;
             case MSG_SMS:
                 String phone = buf.readBytes(21).toString(StandardCharsets.US_ASCII);
-                String realnumber = phone.replaceAll("^(.*)\\+", "+").trim();
+                String sanitizedPhone = phone.substring(phone.indexOf("+")).trim();
                 int remainingbytes = buf.readableBytes();
                 String message = buf.readBytes(remainingbytes).toString(StandardCharsets.UTF_8);
-                position.set(Position.KEY_RESULT, "From: " + realnumber + "\n" + message);
+                position.set("phone_no", sanitizedPhone);
+                position.set(Position.KEY_RESULT, message);
 
-                String responsemessage;
+                String responseMessage;
 
                 switch (message.trim().toUpperCase()) {
                     case "PING?":
                     case "PING#":
-                        responsemessage = "PONG";
+                        responseMessage = "PONG";
                         break;
                     default:
-                        responsemessage = "";
+                        responseMessage = null;
                         break;
                 }
 
-                // pad number to 21 bytes
-                byte[] pp = phone.getBytes(StandardCharsets.US_ASCII);
-                ChannelBuffer ppp = ChannelBuffers.buffer(21);
-                ppp.writeBytes(pp);
-                for (int j = pp.length; j < 21; j++) {
-                    ppp.writeByte(0x00);
-                }
+                // send back response
+                if (responseMessage != null && !responseMessage.isEmpty()) {
+                    // pad number to 21 bytes
+                    ChannelBuffer paddedPhone = ChannelBuffers.buffer(21);
+                    paddedPhone.writeBytes(phone.getBytes(StandardCharsets.US_ASCII));
+                    for (int j = paddedPhone.readableBytes(); j < 21; j++) {
```

Comment:
capacity() - will tell me its 21 bytes long, readableBytes() tells me how many bytes has been written so far.

Are you saying that putting writerIndex = capacity() it should essentially zero pad vs me doing the for loop? If thats the case then yes I would guess it works. Ile test.